### PR TITLE
Issue #19: AreTomo plugin WIP

### LINKED ISSUE #19: AreTomo feature
Add feature to run AreTomo in O2R

### PR DESCRIPTION
**Pull request by elainehoml**
# Description

The [AreTomo](https://www.biorxiv.org/content/10.1101/2022.02.15.480593v1) for O2R allows users to perform alignment, reconstruction, or both on their data. 

The three supported workflows will be
- Alignment
- Reconstruction (WBP or SART)
- Alignment + Reconstruction (WBP or SART)

Additionally, the following corrections will be applied with the use of kwargs into the `ArgParse` in `main/create_aretomo_yaml`
- Dose correction (requires pixel size, dose per tilt, tilt collection scheme, kV
- CTF correction (requires spherical aberration)
- Defocus correction (requires defocus)

Fixes #19  

# Usage

Most basic usage (alignment only, defaults for everything)
`o2r.aretomo.new <proj_name> 0 <rot_angle>`
`o2r.aretomo.run <proj_name> 0`

Recon only
`o2r.aretomo.new <proj_name> 1 <rot_angle> ...` and other parameters as necessary for reconstruction
`o2r.aretomo.run <proj_name> 1`

Align + recon (intermediate alignment step not saved)
`o2r.aretomo.new <proj_name> 2 <rot_angle> ...` and other parameters as necessary for reconstruction
`o2r.aretomo.run <proj_name> 2`

Note that the yaml file for each process is saved under a different filename:
- alignment only (`aretomo_mode=0`) = `<project_name>_aretomo_align.yaml`
- recon only (`aretomo_mode=1`) = `<project_name>_aretomo_recon.yaml`
- alignment + recon (`aretomo_mode=2`) = `<project_name>_aretomo_align-recon.yaml`

# Questions

- [x] 1. Can we repurpose existing code to prepare data for alignment? AreTomo requires an input mrc as a stack (motioncor2 output for alignment workflow, or IMOD/equivalent output for reconstruction workflow), and tilt angle file (raw tilts for alignment workflow, or corrected tilts for reconstruction? _**Addressed in PR #25**_

- [x] 2. The height of the reconstructed volume `VolZ` is by default automatically set by the formula VolZ = Sample thickness * Pixel size + 200 (taken from documentation), though this can be set as an absolute value if necessary. Is it feasible to ask the user to input the sample thickness/pixel size themselves? Can the pixel size be extracted from the mrc metadata somehow? _**Left as user input for now**_

# To-do

- [x] Creation of AreTomo yamls
- [x] AreTomo align workflow
- [x] AreTomo recon workflow
- [x] AreTomo align+recon workflow
- [x] Tests

### CODE REVIEW

File: src/Ot2Rec/main.py [hunk 1]
```diff
@@ -29,16 +29,16 @@
...
-from . import metadata as mdMod
-from . import motioncorr as mc2Mod
-from . import logger as logMod
-from . import ctffind as ctfMod
-from . import align as alignMod
-from . import recon as reconMod
-from . import ctfsim as ctfsimMod
-from . import savurecon as savuMod
-from . import rlf_deconv as rlfMod
+from Ot2Rec import params as prmMod
```
Comment (elainehoml) [inline src/Ot2Rec/main.py:32]: I've changed this import because my IDE was complaining about it

File: src/Ot2Rec/main.py [hunk 2]
```diff
@@ -978,6 +977,90 @@ def run_recon():
         recon_obj.recon_stack()
 
 
+def update_savurecon_yaml(args):
```
Comment (elainehoml) [inline src/Ot2Rec/main.py:980]: Rearranged some functions into a more sensible order

File: src/Ot2Rec/main.py [hunk 3]
```diff
@@ -1361,3 +1290,218 @@ def run_rlf_deconv():
...
+        project_name=args.project_name,
+        filename=aretomo_yaml_name
+    )
+
+    # Add optional kwargs
+    for param in kwargs:
+        aretomo_params.params["AreTomo_kwargs"][param] = vars(args).get(param)
+
+
+    # set process list, InMrc, OutMrc, AngFile
```
Comment (elainehoml) [inline src/Ot2Rec/main.py:1387]: this is where the pre-processing outputs would be required:
- `InMrc` is the input mrc file: either motion-corrected or aligned single .mrc stack
- `OutMrc` is the filename to write the output mrc file
- `AngFile` is the raw or aligned tilt angles data.

File: src/Ot2Rec/main.py [hunk 4]
```diff
@@ -1361,3 +1290,218 @@ def run_rlf_deconv():
...
+
+    # for workflows with reconstruction, set VolZ unless already overwritten
+    if args.aretomo_mode > 0:
+        if args.volz == -1:
+            if args.sample_thickness < 0:
+                raise ValueError("Please set sample thickness in nm to automatically calculate VolZ")
+            if args.pixel_size < 0:
+                raise ValueError("Please set pixel size in nm to automatically calculate VolZ")
+            aretomo_params.params["AreTomo_recon"]["volz"] = int(
+                (args.sample_thickness * args.pixel_size) + 200)
```
Comment (elainehoml) [inline src/Ot2Rec/main.py:1405]: Is there a place where we can extract pixel size automatically from metadata?

File: src/Ot2Rec/main.py [hunk 5]
```diff
@@ -1361,3 +1290,218 @@ def run_rlf_deconv():
...
+    parser.add_argument("--recon_algo",
+                        type=int,
+                        default=0,
+                        help=
+                            "Reconstruction algorithm to be used."
+                            " 0: WBP (default),"
+                            " 1: SART")
+    
+    # Slightly hacky way of adding optional arbitrary arguments, like **kwargs in argparse
+    parsed, unknown = parser.parse_known_args()
```
Comment (elainehoml) [inline src/Ot2Rec/main.py:1491]: There are a lot of kwargs that can be passed to AreTomo, and to keep O2R light I've set up the argparser to take in any key-value pair that will be passed to AreTomo. This is intended for the corrections that can be optionally applied (dose, CTF, defocus) but possibly their extra functionality, e.g., flipping the order of axes, patch-tracking settings etc. Will need to be documented properly though

**Review by elainehoml — COMMENTED** [review]
Still need to get the input mrc and tilt angle files before proceeding to generate AreTomo commands

**Comment by elainehoml** [comment]
> Answer to Q2:
> 
> Yes definitely - I can think of two ways of doing that:
> 
> 1. Import the mrcfile library and use it to read in the stack file (as numpy ndarray). The pixel spacing and other quantities in the metadata will be attributes of the image object.

This is convenient if the motion corrected or aligned tilt series is in mrc, but wouldn't work for tiffs or EER. Perhaps we might need to have a converter class to go between the formats, or convert everything to one standard for the pipeline, e.g. mrc.
Does the EER input option convert to other file formats?

> 2. Manually (using subprocess.run) to call IMOD header and grep the pixel spacing from the stdout
> 
> ....I guess option 1 would be much easier? I used 2 for the EER support only because it's the only way (that I can think of) to read the metadata from the EER files.

Hmm, a bit less straightforward but if you have already implemented this maybe we could have a utils class where we do practical things like extracting any parameter from an IMOD header?

**Comment by elainehoml** [comment]
> I guess we could...? Currently it's only one-purpose (for extracting the number of movie frames from EER files) so everything is hard-coded and very specific. But I guess because the headers for TIFFs / MRCs / EERs are different and their representations are different in the metadata, it might be rather difficult to generalise things.
> 
> Are there specific parameters you need to extract (regardless of image filetypes)?

More generally, usually image dimensions and bit depth is useful. For AreTomo there's spherical aberration, defocus, kV, dose per tilt, tilt collection scheme, but I'm not sure if these are in the headers or not? Maybe have the option to search for these parameters in the headers and then have users define it if not found. In the meantime maybe we can leave it as user-defined parameters, and I can open a new issue for us to set this up later after having a more thorough think about how we can do this and what we can get out of the headers.

**Comment by MasterVexillen** [comment]
The quantities (Cs, delta-Z, voltage etc) you listed are microscope settings so will be in the mdoc files. I don't expect to see many of them in the image headers...

I have a function in metadata for extracting the dosage from the mdoc file.

File: tests/test_argparse.py [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+        self.assertEqual(args.gpu_mem_usage, 1)
+        self.assertEqual(args.exec_path, '/opt/lmod/modules/motioncor2/1.4.0/MotionCor2_1.4.0/MotionCor2_1.4.0_Cuda110')
+        self.assertEqual(args.gain, None)
+        self.assertFalse(args.super_res)
+        self.assertEqual(args.discard_top, 0)
+        self.assertEqual(args.discard_bottom, 0)
+        self.assertEqual(args.tolerance, 0.5)
+        self.assertEqual(args.max_iter, 10)
+        self.assertEqual(args.patch_size, [5, 5, 20])
+        self.assertTrue(args.no_subgroups)
```
Comment (MasterVexillen) [inline tests/test_argparse.py:70]: This one should be assertFalse

File: tests/test_argparse.py [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+        self.assertEqual(args.num_iter, 4)
+        self.assertEqual(args.limits_on_shift, [2, 2])
+        self.assertFalse(args.adjust_tilt_angles)
+        self.assertEqual(args.num_surfaces, 1)
+        self.assertEqual(args.mag_option, 'fixed')
+        self.assertEqual(args.tilt_option, 'fixed')
+        self.assertEqual(args.rot_option, 'group')
+        self.assertEqual(args.beam_tilt_option, 'fixed')
+        self.assertTrue(args.no_robust_fitting)
+        self.assertTrue(args.no_weight_contours)
```
Comment (MasterVexillen) [inline tests/test_argparse.py:128]: These should be assertFalse as well

File: tests/test_argparse.py [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+        self.assertEqual(args.project_name, "test")
+        self.assertEqual(args.unbinned_thickness, 1500)
+        self.assertEqual(args.thickness, 1500)
+
+        self.assertFalse(args.do_positioning)
+        self.assertFalse(args.correct_ctf)
+        self.assertFalse(args.erase_gold)
+        self.assertFalse(args.filtering)
+        self.assertEqual(args.bin_factor, 1)
+        self.assertTrue(args.no_trimvol)
```
Comment (MasterVexillen) [inline tests/test_argparse.py:162]: This one should be assertFalse

**Pull request merged**